File: backend/app/schemas/audit.py
from datetime import datetime
from typing import Any
from uuid import UUID

from app.schemas.common import BaseSchema


class AuditLogResponse(BaseSchema):
    id: UUID
    org_id: UUID
    user_id: UUID | None = None
    deal_id: UUID | None = None
    action: str
    resource_type: str
    resource_id: UUID | None = None
    details: dict[str, Any] | None = None
    ip_address: str | None = None
    created_at: datetime


class AuditLogQuery(BaseSchema):
    user_id: UUID | None = None
    deal_id: UUID | None = None
    action: str | None = None
    resource_type: str | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None

File: worker/internal/store/audit.go
package store

import (
	"context"
	"encoding/json"

	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/util"
)

// Audit is one audit_logs row (ports app/db/audit.record_audit).
type Audit struct {
	OrgID        string
	UserID       *string
	DealID       *string
	Action       string
	ResourceType string
	ResourceID   *string
	Details      any // marshalled to the JSON column; nil -> NULL
}

// RecordAudit appends an audit_logs row. Best-effort callers may ignore the
// error, but writes that are part of a tenant action should propagate it.
func RecordAudit(ctx context.Context, q execer, a Audit) error {
	var details any
	if a.Details != nil {
		b, err := json.Marshal(a.Details)
		if err != nil {
			return err
		}
		details = string(b)
	}
	_, err := q.ExecContext(ctx,
		`INSERT INTO audit_logs(id, org_id, user_id, deal_id, action, resource_type, resource_id, details, created_at)
		 VALUES (?,?,?,?,?,?,?,?,?)`,
		util.NewUUID(), a.OrgID, a.UserID, a.DealID, a.Action, a.ResourceType, a.ResourceID, details, util.NowISO())
	return err
}
